From f87d77d88b820b28c7a3492910c2d9ba5caf26cc Mon Sep 17 00:00:00 2001
Message-Id: <f87d77d88b820b28c7a3492910c2d9ba5caf26cc.1421897934.git.chang-joon.lee@intel.com>
In-Reply-To: <149a6863e085e113eb3c4f75ec994895411447d6.1421897934.git.chang-joon.lee@intel.com>
References: <149a6863e085e113eb3c4f75ec994895411447d6.1421897934.git.chang-joon.lee@intel.com>
From: Namrta <namrta.salonie@intel.com>
Date: Thu, 18 Dec 2014 00:22:48 +0530
Subject: [PATCH 6/7] FOR_UPSTREAM [VPG]: i915/drm: Fix possible null
 dereference in pipe & fb code

Found by using static code analysis tools

v2: Fix commit msg & indentation in code (Rafael)

v3: Fix upper bound check (David)

Issue: GMINL-4362
Change-Id: I2c2229617f9c5e1eeebf6ff96c9a5df4007f2d84
Signed-off-by: Namrta <namrta.salonie@intel.com>
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/i915_debugfs.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_debugfs.c b/drivers/gpu/drm/i915/i915_debugfs.c
index a4d4bd6..2ffd225 100644
--- a/drivers/gpu/drm/i915/i915_debugfs.c
+++ b/drivers/gpu/drm/i915/i915_debugfs.c
@@ -1897,6 +1897,11 @@ static int i915_gem_framebuffer_info(struct seq_file *m, void *data)
 		return ret;
 
 	ifbdev = dev_priv->fbdev;
+	if (!ifbdev) {
+		mutex_unlock(&dev->mode_config.mutex);
+		return -ENOMEM;
+	}
+
 	fb = to_intel_framebuffer(ifbdev->helper.fb);
 
 	seq_printf(m, "fbcon size: %d x %d, depth %d, %d bpp, refcount %d, obj ",
@@ -3331,10 +3336,15 @@ static int pipe_crc_set_source(struct drm_device *dev, enum pipe pipe,
 			       enum intel_pipe_crc_source source)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
-	struct intel_pipe_crc *pipe_crc = &dev_priv->pipe_crc[pipe];
+	struct intel_pipe_crc *pipe_crc = NULL;
 	u32 val = 0; /* shut up gcc */
 	int ret;
 
+	if ((pipe < PIPE_A) || (pipe >= I915_MAX_PIPES))
+		return -EINVAL;
+
+	pipe_crc = &dev_priv->pipe_crc[pipe];
+
 	if (pipe_crc->source == source)
 		return 0;
 
-- 
1.7.9.5

